src/crypto/index.js

```diff
@@ -328,39 +387,49 @@ Crypto.prototype.checkPrivateKey = function(privateKey, expectedPublicKey) {
     }
 };
 
-/**
- * Check whether we already have cross-signing keys for the current user.
- *
- * @return {boolean} Whether we have keys.
- */
-Crypto.prototype.doesCrossSigningHaveKeys = function() {
-    return this._crossSigningInfo.hasKeys();
-};
-
 /**
  * Generate new cross-signing keys.
  *
- * @param {object} authDict Auth data to supply for User-Interactive auth.
  * @param {CrossSigningLevel} [level] the level of cross-signing to reset.  New
  * keys will be created for the given level and below.  Defaults to
  * regenerating all keys.
- */
-Crypto.prototype.resetCrossSigningKeys = async function(authDict, level) {
-    await this._crossSigningInfo.resetKeys(level);
-    await this._signObject(this._crossSigningInfo.keys.master);
-    await this._cryptoStore.doTxn(
-        'readwrite', [IndexedDBCryptoStore.STORE_ACCOUNT],
-        (txn) => {
-            this._cryptoStore.storeCrossSigningKeys(txn, this._crossSigningInfo.keys);
-        },
-    );
+ * @param {function} [opts.doInteractiveAuthFlow] Optional. Function called to
+ * await an interactive auth request.
+ * Args:
+ *     {function} A function that makes the request requiring auth. Receives the
+ *     auth data as an object.
+ */
+Crypto.prototype.resetCrossSigningKeys = async function(level, {
+    doInteractiveAuthFlow = async func => await func(),
+} = {}) {
+    logger.info(`Resetting cross-signing keys at level ${level}`);
+    try {
+        await this._crossSigningInfo.resetKeys(level);
+        await this._signObject(this._crossSigningInfo.keys.master);
+
+        // send keys to server first before caching locally (to ensure upload succeeds)
+        const keys = {};
+        for (const [name, key] of Object.entries(this._crossSigningInfo.keys)) {
+            keys[name + "_key"] = key;
+        }
+        await doInteractiveAuthFlow(async authDict => {
+            await this._baseApis.uploadDeviceSigningKeys(authDict, keys);
+        });
```

Comment:
We could also take a copy of the old keys (ie. probably the empty object) and set them back on the CrossSigningInfo object if this throws, so everything's back in the state it started if the user aborts the UI auth flow.

Reply:
Good idea, added.